src/main/java/net/neoforged/neoforge/event/entity/player/UseItemOnBlockEvent.java

```diff
@@ -0,0 +1,102 @@
+package net.neoforged.neoforge.event.entity.player;
+
+import net.minecraft.world.InteractionResult;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.context.UseOnContext;
+import net.minecraft.world.level.block.state.BlockBehaviour;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
+import net.neoforged.neoforge.common.extensions.IItemExtension;
+
+/**
+ * <p></p>Fires on both the client and server thread when a player uses an item targeting a block.
+ *
+ * <p>The event fires in three phases, corresponding with the three interaction behaviors:
+ * {@link IItemExtension#onItemUseFirst},
+ * {@link BlockBehaviour#use},
+ * and {@link Item#useOn}.</p>
+ *
+ * <p>If the event is cancelled via {@link #cancelWithResult},
+ * then the normal interaction behavior for that phase will not run,
+ * and the specified {@link InteractionResult} will be returned instead.</p>
+ */
+public class UseItemOnBlockEvent extends Event implements ICancellableEvent
+{
+    private final UseOnContext context;
+    private final UsePhase usePhase;
+    private InteractionResult cancellationResult = InteractionResult.PASS;
+
+    public UseItemOnBlockEvent(UseOnContext context, UsePhase usePhase)
+    {
+        this.context = context;
+        this.usePhase = usePhase;
+    }
+
+    /**
+     * @return context
+     */
+    public UseOnContext getUseOnContext()
```

Comment:
It might be worth including a couple helpers to bounce the getters from `UseOnContext` directly into the event - https://github.com/Shadows-of-Fire/Placebo/blob/1.20/src/main/java/dev/shadowsoffire/placebo/events/ItemUseEvent.java#L51-L99
This makes it simpler to access relevant context without always needing to assign the use context to a local before doing anything, and also makes code more portable between events (which it is likely that many people will be doing).

Reply:
I'd rather not, because in that case we might as well just make it extend PlayerInteractEvent, which I was trying to avoid doing for two reasons
* I like event types being flatter and less hierarchical
* Currently all the subclasses of PlayerInteractEvent are static classes in the same file, with names like EntityInteract and RightClickBlock, which I don't like because they're harder to find and they don't have the word "Event" in the name

Reply:
We can utilize extension without making it a static innerclass - I think the innerclass pattern was voted on to be something we should move away from in general.  `PlayerInteractEvent` should probably be made abstract, as well, since it isn't a fired event (and abstract events can no longer be listened to).

Reply:
Fair, made it extend PlayerInteractEvent so it inherits all those getters